rs/replicated_state/src/canister_state/system_state/wasm_chunk_store.rs

```diff
@@ -135,15 +135,26 @@ impl WasmChunkStore {
         Ok(())
     }
 
-    pub fn insert_chunk(&mut self, max_size: NumBytes, chunk: &[u8]) -> Result<[u8; 32], String> {
-        let hash = ic_crypto_sha2::Sha256::hash(chunk);
+    pub fn contains_chunk(&self, hash: &[u8; 32]) -> bool {
+        self.metadata.chunks.contains_key(hash)
+    }
+
+    pub fn insert_chunk(&mut self, max_size: NumBytes, chunk: &[u8], hash: &[u8; 32]) {
+        debug_assert!(
+            *hash == ic_crypto_sha2::Sha256::hash(chunk),
+            "Provided hash {:?} does not match actual chunk hash {:?}",
+            hash,
+            ic_crypto_sha2::Sha256::hash(chunk)
+        );
 
         // No changes needed if we already have the chunk
-        if self.metadata.chunks.contains_key(&hash) {
```

Comment:
Wasn't `upload_chunk` already idempotent, given this check? I'm probably missing something?

Reply:
Not really, because before the function `insert_chunk` is invoked by the management canister implementation, the management canister performs checks (e.g., available memory) and updates (e.g., reserving cycles for extra memory usage) as if the chunk was not available and thus supposed to actually be inserted: [all this](https://github.com/dfinity/ic/blob/53ac5a6a116e58e1530354ea78d6feb5926265a0/rs/execution_environment/src/canister_manager.rs#L1191-L1266) would be executed before realizing we already have the chunk.